### PR TITLE
Rule Concurrency: Prevent flapping of concurrency

### PR DESCRIPTION
**Pull request by julienduchesne**
#### What this PR does

Iterates on https://github.com/grafana/mimir/pull/8146

The `isGroupAtRisk` function only uses the group's last evaluation time as a metric 
However, if the concurrency of the group causes the group's eval time to lower to less than the threshold, this will flap between enabling concurrency and disabling it on every run

In this PR, a condition is added to also sum up the last evaluation time of each rule to compare against the threshold

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/ruler/rule_concurrency.go
Comment (gotjosh) [inline pkg/ruler/rule_concurrency.go:201]:
```suggestion
	for _, r := range group.Rules() {
```

or something similar to avoid shadowing the rule package.

**Review by gotjosh — COMMENTED** [review]
This looks good given it fixes the immediately problem -- However, I'd prefer if we have 2 fields for the runtime - one without concurrency and one with concurrency that we'd only update once the group finishes evaluating. 

I think this will make the runtimes more deterministic and it'll also help us see the difference between expected runtime vs actual runtime (if we expose it as a metric).

The complex part is that we need to do that in Prometheus.

**Comment by gotjosh** [comment]
We also discussed offline, and we're going to implement this change directly in Prometheus. We agreed that we should two things:

1. Keep an atomic with the total runtime of the rules of the group (not how long the group actually took the evaluate) then we can use that directly in order to determine wether this group is eligible for concurrency or not.
2. Add a gauge pretty similar to `prometheus_rule_group_last_duration_seconds` so that we can actually tell when a rule is taking advantage of concurrency.

**Comment by gotjosh** [comment]
Can you please add a changelog entry and make sure you map the newly introduced metric in https://github.com/grafana/mimir/blob/main/pkg/ruler/manager_metrics.go?

**Comment by julienduchesne** [comment]
> Can you please add a changelog entry and make sure you map the newly introduced metric in https://github.com/grafana/mimir/blob/main/pkg/ruler/manager_metrics.go?

Done!

File: pkg/ruler/manager_metrics.go
Comment (gotjosh) [inline pkg/ruler/manager_metrics.go:167]:
```suggestion
	data.SendSumOfGaugesPerTenant(out, m.GroupLastRuleDurationSum, "prometheus_rule_group_last_rule_duration_sum_seconds", dskit_metrics.WithLabels("rule_group"))
```

Comment (gotjosh) [reply pkg/ruler/manager_metrics.go:167]: You need the original metric name here so that we can gather the data from it - you've already renamed it in the collector up top.

File: CHANGELOG.md
Comment (gotjosh) [inline CHANGELOG.md:7]:
```suggestion
* [ENHANCEMENT] Ruler: Add `cortex_prometheus_rule_group_last_rule_duration_sum_seconds` metric to track the total evalation duration of a rule group regardless of concurrency #10189
```

Comment (gotjosh) [reply CHANGELOG.md:7]: Technically an enhancement.

File: CHANGELOG.md
Comment (gotjosh) [inline CHANGELOG.md:8]:
```suggestion
* [BUGFIX] Ruler: Fixed an issue when using the experimental `-ruler.max-independent-rule-evaluation-concurrency` feature, where if a rule group was eligible for concurrency, it would flap between running concurrently or not based on the time it took after running concurrently. #9726 #10189
```

File: go.mod
Comment (gotjosh) [inline go.mod:289]: Please remember to update mimir prometheus before merging this PR - update it in a separate PR.

**Review by gotjosh — APPROVED** [review]
LGTM

Approving so that you don't need another review from me, please make sure you update mimir prometheus before merging this PR.

File: CHANGELOG.md
Comment (tacole02) [inline CHANGELOG.md:13]:
```suggestion
* [ENHANCEMENT] Ruler: Add `cortex_prometheus_rule_group_last_rule_duration_sum_seconds` metric to track the total evaluation duration of a rule group regardless of concurrency #10189
```

Comment (julienduchesne) [reply CHANGELOG.md]: Thanks!

**Pull request merged**
